File: engine/status.py
# Status condition logic: application, prevention, end-of-turn effects

from __future__ import annotations

import random as _random
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .pokemon import Pokemon


# ============================================================
# STATUS CONSTANTS
# ============================================================

BRN = "brn"
PAR = "par"
SLP = "slp"
FRZ = "frz"
PSN = "psn"
TOX = "tox"

NON_VOLATILE = {BRN, PAR, SLP, FRZ, PSN, TOX}

# ailment_id mapping from move data
AILMENT_MAP = {
    1: PAR,
    2: SLP,
    3: FRZ,
    4: BRN,
    5: PSN,  # Toxic overrides to TOX by move name
}

# type immunities: type -> set of statuses it's immune to
TYPE_IMMUNITIES: dict[str, set[str]] = {
    "fire": {BRN},
    "electric": {PAR},
    "poison": {PSN, TOX},
    "steel": {PSN, TOX},
    "ice": {FRZ},
}


# ============================================================
# APPLICATION
# ============================================================

def can_apply_status(pokemon: Pokemon, status: str) -> tuple[bool, str]:
    """Check if a status can be applied. Returns (can_apply, reason)."""
    # already has a non-volatile status
    if pokemon.status is not None:
        return False, "already has a status"

    if pokemon.is_fainted:
        return False, "fainted"

    # type immunities
    for ptype in pokemon.types:
        if ptype in TYPE_IMMUNITIES and status in TYPE_IMMUNITIES[ptype]:
            return False, f"{ptype} type is immune"

    return True, ""


def apply_status(
    pokemon: Pokemon,
    status: str,
    rng: _random.Random,
) -> bool:
    """Apply a non-volatile status. Returns True if applied."""
    can, _ = can_apply_status(pokemon, status)
    if not can:
        return False

    pokemon.status = status

    if status == SLP:
        # gen 2: sleep lasts 1-7 turns
        pokemon.status_turns = rng.randint(1, 7)
    elif status == TOX:
        # toxic counter starts at 1 (1/16 first turn)
        pokemon.status_turns = 0
    else:
        pokemon.status_turns = 0

    return True


def apply_confusion(pokemon: Pokemon, rng: _random.Random) -> bool:
    """Apply confusion. Returns True if applied."""
    if pokemon.confusion_turns > 0 or pokemon.is_fainted:
        return False
    # gen 2: confusion lasts 2-5 turns
    pokemon.confusion_turns = rng.randint(2, 5)
    return True


# ============================================================
# MOVE PREVENTION
# ============================================================

def check_move_prevention(
    pokemon: Pokemon,
    rng: _random.Random,
) -> tuple[bool, str | None]:
    """
    Check if a pokemon can act this turn due to status.
    Returns (can_act, reason_string_or_None).
    """
    status = pokemon.status

    if status == SLP:
        if pokemon.status_turns > 0:
            pokemon.status_turns -= 1
            if pokemon.status_turns == 0:
                pokemon.clear_status()
                return True, "woke up"
            return False, "fast asleep"
        # counter hit 0 -- wake up
        pokemon.clear_status()
        return True, "woke up"

    if status == FRZ:
        # 20% chance to thaw each turn
        if rng.random() < 0.20:
            pokemon.clear_status()
            return True, "thawed out"
        return False, "frozen solid"

    if status == PAR:
        # 25% chance to be fully paralyzed
        if rng.random() < 0.25:
            return False, "fully paralyzed"

    return True, None


def check_confusion(
    pokemon: Pokemon,
    rng: _random.Random,
) -> tuple[bool, int]:
    """
    Check confusion before attacking. Returns (hit_self, self_damage).
    Decrements confusion counter. If hit_self is True, the pokemon
    hits itself for self_damage instead of using its move.
    """
    if pokemon.confusion_turns <= 0:
        return False, 0

    pokemon.confusion_turns -= 1

    # 50% chance to hit self
    if rng.random() < 0.5:
        # 40 power typeless physical hit against self
        level = 100
        atk = pokemon.stats["attack"]
        dfn = pokemon.stats["defense"]
        # simplified damage formula: no stab, no type, no crit, no random
        damage = ((2 * level // 5 + 2) * 40 * atk // dfn) // 50 + 2
        return True, damage

    return False, 0


# ============================================================
# END-OF-TURN EFFECTS
# ============================================================

def end_of_turn_damage(pokemon: Pokemon) -> int:
    """Calculate end-of-turn residual damage. Returns damage amount (0 if none)."""
    status = pokemon.status

    if status == BRN:
        return pokemon.max_hp // 8

    if status == PSN:
        return pokemon.max_hp // 8

    if status == TOX:
        pokemon.status_turns += 1
        return (pokemon.max_hp * pokemon.status_turns) // 16

    return 0


# ============================================================
# STATUS FROM MOVE DATA
# ============================================================

def status_from_move(move_name: str, ailment_id: int) -> str | None:
    """Determine which status a move applies. Returns None if out of scope."""
    if ailment_id not in AILMENT_MAP:
        return None

    # Toxic (move id 92) applies toxic poison, not regular poison
    if ailment_id == 5 and move_name == "Toxic":
        return TOX

    return AILMENT_MAP[ailment_id]


def confusion_from_move(ailment_id: int) -> bool:
    """Check if a move applies confusion."""
    return ailment_id == 6


def effective_speed(pokemon: Pokemon) -> int:
    """Get effective speed accounting for stat stages and paralysis."""
    from .stat_stages import get_stage_multiplier
    speed = pokemon.stats["speed"]
    stage = pokemon.stat_stages.get("speed", 0)
    if stage != 0:
        num, den = get_stage_multiplier(stage)
        speed = speed * num // den
    if pokemon.status == PAR:
        speed //= 4
    return speed

File: crystal_engine/src/status.rs
// Status effects: application, prevention, and end-of-turn damage

use rand::Rng;

use crate::pokemon::Pokemon;
use crate::stat_stages::get_stage_multiplier;
use crate::types::Type;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Status {
    None,
    Burn,
    Paralysis,
    Sleep,
    Freeze,
    Poison,
    Toxic,
}

impl Status {
    pub fn is_none(self) -> bool {
        self == Status::None
    }

    pub fn is_some(self) -> bool {
        self != Status::None
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Status::None => "",
            Status::Burn => "brn",
            Status::Paralysis => "par",
            Status::Sleep => "slp",
            Status::Freeze => "frz",
            Status::Poison => "psn",
            Status::Toxic => "tox",
        }
    }

    pub fn from_str(s: &str) -> Status {
        match s {
            "brn" => Status::Burn,
            "par" => Status::Paralysis,
            "slp" => Status::Sleep,
            "frz" => Status::Freeze,
            "psn" => Status::Poison,
            "tox" => Status::Toxic,
            _ => Status::None,
        }
    }
}

/// Check if a status can be applied to a pokemon. Returns (can_apply, reason).
pub fn can_apply_status(pokemon: &Pokemon, status: Status) -> (bool, &'static str) {
    if pokemon.status.is_some() {
        return (false, "already has status");
    }
    if pokemon.is_fainted() {
        return (false, "fainted");
    }
    // type immunities
    for &t in pokemon.types() {
        match t {
            Type::Fire if status == Status::Burn => return (false, "fire immune to burn"),
            Type::Electric if status == Status::Paralysis => {
                return (false, "electric immune to paralysis")
            }
            Type::Poison | Type::Steel
                if status == Status::Poison || status == Status::Toxic =>
            {
                return (false, "immune to poison")
            }
            Type::Ice if status == Status::Freeze => return (false, "ice immune to freeze"),
            _ => {}
        }
    }
    (true, "")
}

/// Apply a status to a pokemon. Returns true if applied.
pub fn apply_status(pokemon: &mut Pokemon, status: Status, rng: &mut impl Rng) -> bool {
    let (can, _) = can_apply_status(pokemon, status);
    if !can {
        return false;
    }
    pokemon.status = status;
    match status {
        Status::Sleep => pokemon.status_turns = rng.random_range(1..=7),
        Status::Toxic => pokemon.status_turns = 0,
        _ => pokemon.status_turns = 0,
    }
    true
}

/// Apply confusion. Returns true if applied.
pub fn apply_confusion(pokemon: &mut Pokemon, rng: &mut impl Rng) -> bool {
    if pokemon.confusion_turns > 0 || pokemon.is_fainted() {
        return false;
    }
    pokemon.confusion_turns = rng.random_range(2..=5);
    true
}

/// Result of checking whether a move is prevented by status
pub enum MovePreventionResult {
    CanAct,
    WokeUp,
    ThawedOut,
    Prevented(&'static str), // reason: "fast asleep", "frozen solid", "fully paralyzed"
}

/// Check if status prevents acting this turn. Mutates pokemon state.
pub fn check_move_prevention(pokemon: &mut Pokemon, rng: &mut impl Rng) -> MovePreventionResult {
    match pokemon.status {
        Status::Sleep => {
            pokemon.status_turns -= 1;
            if pokemon.status_turns == 0 {
                pokemon.clear_status();
                MovePreventionResult::WokeUp
            } else {
                MovePreventionResult::Prevented("fast asleep")
            }
        }
        Status::Freeze => {
            if rng.random_range(0..100) < 20 {
                pokemon.clear_status();
                MovePreventionResult::ThawedOut
            } else {
                MovePreventionResult::Prevented("frozen solid")
            }
        }
        Status::Paralysis => {
            if rng.random_range(0..100) < 25 {
                MovePreventionResult::Prevented("fully paralyzed")
            } else {
                MovePreventionResult::CanAct
            }
        }
        _ => MovePreventionResult::CanAct,
    }
}

/// Check confusion. Returns (hit_self, self_damage).
pub fn check_confusion(pokemon: &mut Pokemon, rng: &mut impl Rng) -> (bool, u16) {
    if pokemon.confusion_turns == 0 {
        return (false, 0);
    }
    pokemon.confusion_turns -= 1;
    // 50% chance to hit self
    if rng.random_range(0..2) == 0 {
        // typeless 40-power physical hit, no crits, no random roll
        let atk = pokemon.stats[1] as u32; // attack
        let dfn = pokemon.stats[2] as u32; // defense
        let damage = ((42 * 40 * atk / dfn) / 50 + 2) as u16;
        (true, damage)
    } else {
        (false, 0)
    }
}

/// Calculate end-of-turn residual damage. Returns damage amount.
/// For toxic, increments status_turns as a side effect.
pub fn end_of_turn_damage(pokemon: &mut Pokemon) -> u16 {
    let max_hp = pokemon.stats[0] as u16;
    match pokemon.status {
        Status::Burn | Status::Poison => max_hp / 8,
        Status::Toxic => {
            pokemon.status_turns += 1;
            (max_hp as u32 * pokemon.status_turns as u32 / 16) as u16
        }
        _ => 0,
    }
}

/// Determine status from move's ailment_id. ailment_id 5 + "Toxic" -> Toxic.
pub fn status_from_move(move_name: &str, ailment_id: u8) -> Option<Status> {
    match ailment_id {
        1 => Some(Status::Paralysis),
        2 => Some(Status::Sleep),
        3 => Some(Status::Freeze),
        4 => Some(Status::Burn),
        5 => {
            if move_name == "Toxic" {
                Some(Status::Toxic)
            } else {
                Some(Status::Poison)
            }
        }
        _ => None,
    }
}

/// Returns true if ailment_id represents confusion
pub fn confusion_from_move(ailment_id: u8) -> bool {
    ailment_id == 6
}

/// Effective speed accounting for stat stages and paralysis
pub fn effective_speed(pokemon: &Pokemon) -> u32 {
    let base_speed = pokemon.stats[5] as u32;
    let stage = pokemon.stat_stages[4]; // speed stage
    let (num, den) = get_stage_multiplier(stage);
    let mut speed = base_speed * num as u32 / den as u32;
    if pokemon.status == Status::Paralysis {
        speed /= 4;
    }
    speed
}
